ldm/deepspeed_replace.py

```diff
@@ -271,37 +273,164 @@ def apply(self, module, enable_cuda_graph=True):
 GENERIC_POLICIES = [UNetPolicy, VAEPolicy, ClipEncoderPolicy]
 
 
+
 def _module_match(module):
     for policy in GENERIC_POLICIES:
         policy = policy()
         if policy.match(module):
             return policy
     return None
 
+
+class FlashAttention(nn.Module):
+
+    def __init__(self, hidden_size: int, hidden_size_out: int, heads: int, fp16: bool = True, device: str = "cuda", max_out_tokens: int = 4096):
+        super().__init__()
+
+        self.hidden_size = hidden_size
+        self.max_out_tokens = max_out_tokens
+        self.heads = heads
+
+        dtype = torch.float16 if fp16 else torch.float32
+
+        self.attn_kw = nn.Parameter(
+            torch.empty(
+                hidden_size,
+                hidden_size,
+                dtype=dtype,
+                device=device
+            ), requires_grad=False
+        )
+
+        self.attn_vw = nn.Parameter(
+            torch.empty(
+                hidden_size,
+                hidden_size,
+                dtype=dtype,
+                device=device
+            ), requires_grad=False)
+
+        self.attn_qw = nn.Parameter(
+            torch.empty(
+                hidden_size,
+                hidden_size_out,
+                dtype=dtype,
+                device=device
+            ), requires_grad=False
+        )
+
+        self.attn_ow = nn.Parameter(torch.empty(hidden_size,
+                                                hidden_size,
+                                                dtype=dtype,
+                                                device=device),
+                                    requires_grad=False)
+
+        self.attn_ob = nn.Parameter(torch.empty(hidden_size,
+                                                dtype=dtype,
+                                                device=device),
+                                    requires_grad=False)
+        self.do_out_bias = True
+
+        self.norm_factor = math.sqrt(math.sqrt(hidden_size // heads))
+
+        self.scale = (1 / self.norm_factor) * (1 / self.norm_factor) 
+
+    def forward(self, input, context=None, input_mask=None):
+        from flash_attn.flash_attn_interface import _flash_attn_forward
+
+        batch_size = input.shape[0]
+        seqlen = input.shape[1]
+        cu_seqlens = torch.arange(0, (batch_size + 1) * seqlen, step=seqlen, dtype=torch.int32, device=self.attn_kw.device)
+        do_flash_attn = (input.shape[-1] / self.heads <= 128) and context is None
+
+        if not do_flash_attn:
```

Comment:
Shouldn't we output some kind of warning here, if the expectation is that this is running with FlashAttention? It kind of breaks the contract.

Reply:
I followed Flash Attention & DeepSpeed internal contract. I didn't change the logic. Unfortunately, flash attention isn't implemented for context and heads higher than 128.

So problably, there are extra speed gains to get there by implementing the kernels.

Reply:
I see, we should make this explicit in the docstrings at the very least, or even better by printing a warning the first time the condition fails.

Reply:
Done !